cmd/prometheus/main.go

```diff
@@ -282,104 +277,193 @@ func main() {
+							rc <- err
+						} else {
+							rc <- nil
+						}
+					case <-cancel:
+						return nil
+					}
+				}
 
-	if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
-		level.Error(logger).Log("msg", "Error loading config", "err", err)
-		os.Exit(1)
+			},
+			func(err error) {
+				close(cancel)
+			},
+		)
 	}
+	{
 
-	defer func() {
-		if err := fanoutStorage.Close(); err != nil {
-			level.Error(logger).Log("msg", "Error stopping storage", "err", err)
-		}
-	}()
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				<-dbOpen
+				if err := reloadConfig(cfg.configFile, logger, reloadables...); err != nil {
+					return fmt.Errorf("Error loading config %s", err)
+				}
 
-	// Wait for reload or termination signals.
-	close(hupReady) // Unblock SIGHUP handler.
-
-	// Set web server to ready.
-	webHandler.Ready()
-	level.Info(logger).Log("msg", "Server is ready to receive requests.")
-
-	term := make(chan os.Signal, 1)
-	signal.Notify(term, os.Interrupt, syscall.SIGTERM)
-	select {
-	case <-term:
-		level.Warn(logger).Log("msg", "Received SIGTERM, exiting gracefully...")
-	case <-webHandler.Quit():
-		level.Warn(logger).Log("msg", "Received termination request via web service, exiting gracefully...")
-	case err := <-errc:
-		level.Error(logger).Log("msg", "Error starting web server, exiting gracefully", "err", err)
+				// Wait for reload or termination signals.
+				close(hupReady) // Unblock SIGHUP handler.
+
+				// Set web server to ready.
+				webHandler.Ready()
+				level.Info(logger).Log("msg", "Server is ready to receive requests.")
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				close(cancel)
+			},
+		)
+	}
+	{
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				db, err := tsdb.Open(
+					cfg.localStoragePath,
+					log.With(logger, "component", "tsdb"),
+					prometheus.DefaultRegisterer,
+					&cfg.tsdb,
+				)
+				if err != nil {
+					return fmt.Errorf("Opening storage failed %s", err)
+				}
+				localStorage.Set(db)
+				close(dbOpen)
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				if err := fanoutStorage.Close(); err != nil {
+					level.Error(logger).Log("msg", "Error stopping storage", "err", err)
+				}
+				close(cancel)
+			},
+		)
+	}
+	{
+		// The notifier is a dependency of the rule manager. It has to be
+		// started before and torn down afterwards.
+		g.Add(
+			func() error {
+				notifier.Run()
+				return nil
+			},
+			func(err error) {
+				notifier.Stop()
+			},
+		)
+	}
+	{
+		//TODO(krasi) refactor ruleManager.Run() to be blocking to avoid using an extra blocking channel
+		cancel := make(chan struct{})
+		g.Add(
+			func() error {
+				ruleManager.Run()
+				<-cancel
+				return nil
+			},
+			func(err error) {
+				ruleManager.Stop()
+				close(cancel)
+			},
+		)
 	}
+	{
+		//TODO(krasi) refactor targetManager.Run() to be blocking to avoid using an extra blocking channel
```

Comment:
Nit: The Prometheus project has been pretty pedantic about following the golang comment style so far. Please leave a space after `//` and end all comments on a dot.

Reply:
ok , I will try to make `targetManager.Run()` blocking so I would remove the comment anyway.